File: src/main/java/org/gangel/graphomance/ObjectApi.java
package org.gangel.graphomance;

import java.util.Map;

public interface ObjectApi {

  void startTransaction();

  void commit();

  void rollback();

  Object getNode(final String clsName, final NodeIdentifier nodeId);

  long shortestPath(final NodeIdentifier start, final NodeIdentifier end);

  default NodeIdentifier createNode(final String clsName) {
    return createNode(clsName, null);
  }

  NodeIdentifier createNode(final String clsName, Map<String, Object> properties);

  void updateNode(final String clsName, NodeIdentifier nodeId, Map<String, Object> properties);

  default RelationIdentifier createRelation(String className, NodeIdentifier fromNode, NodeIdentifier toNode) {
    return createRelation(className, fromNode, toNode, null);
  }

  RelationIdentifier createRelation(String className, NodeIdentifier fromNode, NodeIdentifier toNode,
                                    Map<String, Object> properties);

  void updateRelation(final String clsName, RelationIdentifier edgeId, Map<String, Object> properties);

  void deleteAllNodes(final String clsName);

  void deleteAllRelations(final String clsName);
}

File: src/main/kotlin/org/gangel/graphomance/api/ObjectApi.kt
package org.gangel.graphomance.api

interface ObjectApi {
    fun startTransaction()
    fun commit()
    fun rollback()
    fun getNode(clsName: String, nodeId: NodeIdentifier): Any?
    fun shortestPath(start: NodeIdentifier, end: NodeIdentifier): Long

    fun createNode(clsName: String): NodeIdentifier {
        return createNode(clsName)
    }

    fun createNode(clsName: String, properties: Map<String, Any> = emptyMap()): NodeIdentifier

    fun updateNode(clsName: String, nodeId: NodeIdentifier, properties: Map<String, Any>)

    fun createRelation(className: String, fromNode: NodeIdentifier, toNode: NodeIdentifier): RelationIdentifier {
        return createRelation(className, fromNode, toNode)
    }

    fun createRelation(className: String, fromNode: NodeIdentifier, toNode: NodeIdentifier, properties: Map<String, Any> = emptyMap()): RelationIdentifier

    fun updateRelation(clsName: String, edgeId: RelationIdentifier, properties: Map<String, Any>)

    fun deleteAllNodes(clsName: String)
    fun deleteAllRelations(clsName: String)
}